### PR TITLE
fix: gateway should not be network address and broadcast address

### LINKED ISSUE #4041: [BUG] Gateway should not be the network address and the broadcast address
### Kube-OVN Version

v1.13.0

### Kubernetes Version

v1.25.3

### Operation-system/Kernel Version

"Ubuntu 22.04.4 LTS"  6.5.0-35-generic

### Description

(e.g.) subnet cidr 192.168.0.0/24
subnet.spec.gateway can be set to 192.168.0.0 or 192.168.0.255

### Steps To Reproduce

(e.g.) set subnet.spec.gateway to 192.168.0.0

### Current Behavior

subnet.spec.gateway can be set to 192.168.0.0

### Expected Behavior

subnet.spec.gateway should be checked before created

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Bug fixes

## Which issue(s) this PR fixes

Fixes #4041

### CODE REVIEW

**Review by bobz965 — APPROVED** [review]

**Comment by bobz965** [comment]
@zhangzujian please help review, thanks!

**Comment by zhangzujian** [comment]
`CIDRContainIP` is widely used in Kube-OVN. I think we should not change its behavior. In stead, we can create a new funciton to validate subnet gateway address.

**Comment by zcq98** [comment]
> `CIDRContainIP` is widely used in Kube-OVN. I think we should not change its behavior. In stead, we can create a new funciton to validate subnet gateway address.

但是不仅是subnet的gateway校验有问题，其它能指定IP的地方都存在可以将ip指定为网络地址/广播地址的问题

**Comment by zhangzujian** [comment]
> > `CIDRContainIP` is widely used in Kube-OVN. I think we should not change its behavior. In stead, we can create a new funciton to validate subnet gateway address.
> 
> 但是不仅是subnet的gateway校验有问题，其它能指定IP的地方都存在可以将ip指定为网络地址/广播地址的问题

那我们就把这些地方找出来逐一修复掉。如果工作量比较大，这个 PR 可以先只修复 gateway 地址。

File: pkg/util/net.go [hunk 1]
```diff
@@ -94,6 +94,14 @@ func FirstIP(subnet string) (string, error) {
 	return BigInt2Ip(ipInt.Add(ipInt, big.NewInt(1))), nil
 }
 
+func NetworkAddress(subnet string) (string, error) {
```
Comment (zhangzujian) [inline pkg/util/net.go:97]: We already have `SubnetNumber` and `SubnetBroadcast`.

Comment (zcq98) [reply pkg/util/net.go]: 已修正，目前只做subnet gateway的校验，webhook中也会通过ValidateSubnet校验gateway

File: pkg/util/validator.go [hunk 2]
```diff
@@ -13,8 +13,8 @@ import (
 )
 
 func ValidateSubnet(subnet kubeovnv1.Subnet) error {
-	if subnet.Spec.Gateway != "" && !CIDRContainIP(subnet.Spec.CIDRBlock, subnet.Spec.Gateway) {
-		return fmt.Errorf(" gateway %s is not in cidr %s", subnet.Spec.Gateway, subnet.Spec.CIDRBlock)
+	if subnet.Spec.Gateway != "" && (!CIDRContainIP(subnet.Spec.CIDRBlock, subnet.Spec.Gateway) || isNetworkOrBroadcast(subnet.Spec.CIDRBlock, subnet.Spec.Gateway)) {
```
Comment (zhangzujian) [inline pkg/util/validator.go:16]: Can we use `ValidatePodCidr` here?

Comment (zcq98) [reply pkg/util/validator.go]: sure，i think It is more appropriate to change the function name ValidatePodCidr to ValidateNetworkBroadcast

File: pkg/util/validator.go [hunk 3]
```diff
@@ -13,8 +13,8 @@ import (
 )
 
 func ValidateSubnet(subnet kubeovnv1.Subnet) error {
-	if subnet.Spec.Gateway != "" && !CIDRContainIP(subnet.Spec.CIDRBlock, subnet.Spec.Gateway) {
-		return fmt.Errorf(" gateway %s is not in cidr %s", subnet.Spec.Gateway, subnet.Spec.CIDRBlock)
+	if subnet.Spec.Gateway != "" && (!CIDRContainIP(subnet.Spec.CIDRBlock, subnet.Spec.Gateway) || ValidateNetworkBroadcast(subnet.Spec.CIDRBlock, subnet.Spec.Gateway) != nil) {
+		return fmt.Errorf("gateway %s is not in cidr %s", subnet.Spec.Gateway, subnet.Spec.CIDRBlock)
```
Comment (zhangzujian) [inline pkg/util/validator.go:17]: We should provide detailed information/log if the gateway is network/broadcast address.

Comment (zhangzujian) [inline pkg/util/validator.go:17]: We should provide detailed information/log if the gateway is network/broadcast address.

**Review by zhangzujian — APPROVED** [review]
LGTM. Can be merged after the workflows succeed.

**Pull request merged**